### PR TITLE
[PIL-965][AV] - Repayments Telephone Preference and Contact Details

### PR DESCRIPTION
**Pull request by aashik-varahan**

### CODE REVIEW

File: app/controllers/repayments/RepaymentsContactByTelephoneController.scala [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import repositories.SessionRepository
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.repayments.RepaymentsContactByTelephoneView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class RepaymentsContactByTelephoneController @Inject() (
+  identify:                  IdentifierAction,
+  val userAnswersConnectors: UserAnswersConnectors,
```
Comment (jassalrichy) [inline app/controllers/repayments/RepaymentsContactByTelephoneController.scala:39]: remove unused UserAnswersConnector DI

File: app/controllers/repayments/RepaymentsTelephoneDetailsController.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import repositories.SessionRepository
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.repayments.RepaymentsTelephoneDetailsView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class RepaymentsTelephoneDetailsController @Inject() (
+  identify:                  IdentifierAction,
+  val userAnswersConnectors: UserAnswersConnectors,
```
Comment (jassalrichy) [inline app/controllers/repayments/RepaymentsTelephoneDetailsController.scala:39]: remove unused UserAnswersConnector DI, and it's import

File: app/controllers/repayments/RepaymentsTelephoneDetailsController.scala [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      identify
+    ) andThen getSessionData andThen requireSessionData).async { implicit request =>
+      request.userAnswers
+        .get(RepaymentsContactNamePage)
+        .map { name =>
+          formProvider(name)
+            .bindFromRequest()
+            .fold(
+              formWithErrors => Future.successful(BadRequest(view(formWithErrors, clientPillar2Id, mode, name))),
+              nominated =>
```
Comment (jassalrichy) [inline app/controllers/repayments/RepaymentsTelephoneDetailsController.scala:87]: rename nominated to telephoneNumber

File: test/controllers/repayments/RepaymentsContactByTelephoneControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+import play.api.test.Helpers._
+import repositories.SessionRepository
+import views.html.repayments.RepaymentsContactByTelephoneView
+
+import scala.concurrent.Future
+
+class RepaymentsContactByTelephoneControllerSpec extends SpecBase {
+
+  val formProvider = new RepaymentsContactByTelephoneFormProvider()
+  val form         = formProvider("ABC Limited")
```
Comment (jassalrichy) [inline test/controllers/repayments/RepaymentsContactByTelephoneControllerSpec.scala:36]: Add annotation type

File: test/controllers/repayments/RepaymentsTelephoneDetailsControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+import play.api.test.Helpers._
+import repositories.SessionRepository
+import views.html.repayments.RepaymentsTelephoneDetailsView
+
+import scala.concurrent.Future
+
+class RepaymentsTelephoneDetailsControllerSpec extends SpecBase {
+
+  val formProvider = new RepaymentsTelephoneDetailsFormProvider()
+  val form         = formProvider("ABC Limited")
```
Comment (jassalrichy) [inline test/controllers/repayments/RepaymentsTelephoneDetailsControllerSpec.scala:36]: Add type annotation

File: test/controllers/repayments/RepaymentsContactByTelephoneControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+  "Repayments Contact By Telephone Controller" when {
```
Comment (jassalrichy) [inline test/controllers/repayments/RepaymentsContactByTelephoneControllerSpec.scala:38]: line 38 - "Repayments Contact By Telephone Controller"

File: test/controllers/repayments/RepaymentsTelephoneDetailsControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+  "Repayments Contact Email Controller" when {
```
Comment (jassalrichy) [inline test/controllers/repayments/RepaymentsTelephoneDetailsControllerSpec.scala:38]: line 38 - "Repayments Telephone Details Controller"

File: test/controllers/repayments/RepaymentsContactEmailControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+import play.api.test.Helpers._
+import repositories.SessionRepository
+import views.html.repayments.RepaymentsContactEmailView
+
+import scala.concurrent.Future
+
+class RepaymentsContactEmailControllerSpec extends SpecBase {
+
+  val formProvider = new RepaymentsContactEmailFormProvider()
+  val form         = formProvider("ABC Limited")
```
Comment (jassalrichy) [inline test/controllers/repayments/RepaymentsContactEmailControllerSpec.scala:36]: Add type annotation

File: test/controllers/repayments/RepaymentsContactNameControllerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+import play.api.test.Helpers._
+import repositories.SessionRepository
+import views.html.repayments.RepaymentsContactNameView
+
+import scala.concurrent.Future
+
+class RepaymentsContactNameControllerSpec extends SpecBase {
+
+  val formProvider = new RepaymentsContactNameFormProvider()
+  val form         = formProvider()
```
Comment (jassalrichy) [inline test/controllers/repayments/RepaymentsContactNameControllerSpec.scala:36]: Add type annotation

File: test/forms/RepaymentsContactByTelephoneFormProviderSpec.scala [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+import forms.behaviours.BooleanFieldBehaviours
+import play.api.data.FormError
+
+class RepaymentsContactByTelephoneFormProviderSpec extends BooleanFieldBehaviours {
+
+  val requiredKey = "repayments.contactByTelephone.error.required"
+  val invalidKey  = "error.boolean"
+
+  val formProvider = new RepaymentsContactByTelephoneFormProvider()
+  val form         = formProvider("test")
```
Comment (jassalrichy) [inline test/forms/RepaymentsContactByTelephoneFormProviderSpec.scala:28]: Add type annotation

File: test/forms/RepaymentsContactEmailFormProviderSpec.scala [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import mapping.Constants
+import play.api.data.FormError
+
+class RepaymentsContactEmailFormProviderSpec extends StringFieldBehaviours {
+
+  val requiredKey       = "repayments.contactEmail.error.contactEmail.required"
+  val lengthKey         = "repayments.contactEmail.error.contactEmail.length"
+  val formatKey         = "repayments.contactEmail.error.contactEmail.format"
+  val contactName       = "ABC Limited"
+  val maxLength         = Constants.MAX_LENGTH_100
```
Comment (jassalrichy) [inline test/forms/RepaymentsContactEmailFormProviderSpec.scala:29]: line 29: Add type annotation

File: test/forms/RepaymentsContactEmailFormProviderSpec.scala [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+import forms.behaviours.StringFieldBehaviours
+import mapping.Constants
+import play.api.data.FormError
+
+class RepaymentsContactEmailFormProviderSpec extends StringFieldBehaviours {
+
+  val requiredKey       = "repayments.contactEmail.error.contactEmail.required"
+  val lengthKey         = "repayments.contactEmail.error.contactEmail.length"
+  val formatKey         = "repayments.contactEmail.error.contactEmail.format"
```
Comment (jassalrichy) [inline test/forms/RepaymentsContactEmailFormProviderSpec.scala:27]: remove unused formatKey

File: test/forms/RepaymentsContactNameFormProviderSpec.scala [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+import forms.behaviours.StringFieldBehaviours
+import mapping.Constants
+import play.api.data.FormError
+
+class RepaymentsContactNameFormProviderSpec extends StringFieldBehaviours {
+
+  val requiredKey = "repayments.contactName.error.required"
+  val lengthKey   = "repayments.contactName.error.length"
+  val maxLength   = Constants.MAX_LENGTH_100
```
Comment (jassalrichy) [inline test/forms/RepaymentsContactNameFormProviderSpec.scala:27]: line 27: Add type annotation

File: test/forms/RepaymentsTelephoneDetailsFormProviderSpec.scala [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+class RepaymentsTelephoneDetailsFormProviderSpec extends StringFieldBehaviours {
+
+  val requiredKey = "repayments.telephoneDetails.error.required"
+  val lengthKey   = "repayments.telephoneDetails.error.length"
+  val formatKey   = "repayments.telephoneDetails.error.format"
+  val maxLength   = 50
+  val formatReg   = """^[A-Z0-9 )/(\-*#+]*$"""
+
+  val formProvider = new RepaymentsTelephoneDetailsFormProvider()
+  val form         = formProvider("test")
```
Comment (jassalrichy) [inline test/forms/RepaymentsTelephoneDetailsFormProviderSpec.scala:31]: line 31: Add type annotation

File: test/forms/RepaymentsContactEmailFormProviderSpec.scala [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  val contactName       = "ABC Limited"
+  val maxLength         = Constants.MAX_LENGTH_100
+  val validEmailAddress = "abc@cba.com"
+  val form              = new RepaymentsContactEmailFormProvider()(contactName)
+
+  ".value" - {
```
Comment (jassalrichy) [inline test/forms/RepaymentsContactEmailFormProviderSpec.scala:33]: line33: ".contactEmail"

File: test/forms/RepaymentsContactNameFormProviderSpec.scala [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+  val form        = new RepaymentsContactNameFormProvider()
+
+  ".value" - {
```
Comment (jassalrichy) [inline test/forms/RepaymentsContactNameFormProviderSpec.scala:30]: line30: ".contactName"

File: test/forms/RepaymentsTelephoneDetailsFormProviderSpec.scala [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+  ".value" - {
```
Comment (jassalrichy) [inline test/forms/RepaymentsTelephoneDetailsFormProviderSpec.scala:33]: line33: ".telephoneNumber"

File: test/navigation/RepaymentNavigatorSpec.scala [hunk 18]
```diff
@@ -35,11 +35,61 @@ class RepaymentNavigatorSpec extends SpecBase {
         navigator.nextPage(UnknownPage, None, NormalMode, UserAnswers("id")) mustBe routes.IndexController.onPageLoad
       }
 
-      "go to under construction page from request refund amount page" in {
+      "must go to next page from request refund amount page" in {
```
Comment (jassalrichy) [inline test/navigation/RepaymentNavigatorSpec.scala:38]: line38: "must go to next page.."? please specifiy the page

File: app/controllers/repayments/RepaymentsContactByTelephoneController.scala (hunk 1 above)
Comment (aashik-varahan) [reply app/controllers/repayments/RepaymentsContactByTelephoneController.scala:39]: Removed

File: app/controllers/repayments/RepaymentsTelephoneDetailsController.scala (hunk 2 above)
Comment (aashik-varahan) [reply app/controllers/repayments/RepaymentsTelephoneDetailsController.scala:39]: Removed

File: app/controllers/repayments/RepaymentsTelephoneDetailsController.scala (hunk 3 above)
Comment (aashik-varahan) [reply app/controllers/repayments/RepaymentsTelephoneDetailsController.scala:87]: Renamed to telephoneNumber

File: test/controllers/repayments/RepaymentsContactByTelephoneControllerSpec.scala (hunk 4 above)
Comment (aashik-varahan) [reply test/controllers/repayments/RepaymentsContactByTelephoneControllerSpec.scala]: Added type annotation

File: test/controllers/repayments/RepaymentsTelephoneDetailsControllerSpec.scala (hunk 5 above)
Comment (aashik-varahan) [reply test/controllers/repayments/RepaymentsTelephoneDetailsControllerSpec.scala]: Added type annotation

File: test/controllers/repayments/RepaymentsContactEmailControllerSpec.scala (hunk 8 above)
Comment (aashik-varahan) [reply test/controllers/repayments/RepaymentsContactEmailControllerSpec.scala]: Added type annotation

File: test/controllers/repayments/RepaymentsContactNameControllerSpec.scala (hunk 9 above)
Comment (aashik-varahan) [reply test/controllers/repayments/RepaymentsContactNameControllerSpec.scala]: Added type annotation

File: test/forms/RepaymentsContactByTelephoneFormProviderSpec.scala (hunk 10 above)
Comment (aashik-varahan) [reply test/forms/RepaymentsContactByTelephoneFormProviderSpec.scala]: Added type annotation

File: test/forms/RepaymentsContactEmailFormProviderSpec.scala (hunk 11 above)
Comment (aashik-varahan) [reply test/forms/RepaymentsContactEmailFormProviderSpec.scala]: Added type annotation

File: test/forms/RepaymentsContactEmailFormProviderSpec.scala (hunk 12 above)
Comment (aashik-varahan) [reply test/forms/RepaymentsContactEmailFormProviderSpec.scala]: Removed formatKey

File: test/forms/RepaymentsContactNameFormProviderSpec.scala (hunk 13 above)
Comment (aashik-varahan) [reply test/forms/RepaymentsContactNameFormProviderSpec.scala]: Added type annotation

File: test/forms/RepaymentsTelephoneDetailsFormProviderSpec.scala (hunk 14 above)
Comment (aashik-varahan) [reply test/forms/RepaymentsTelephoneDetailsFormProviderSpec.scala]: Added type annotation

File: test/forms/RepaymentsContactEmailFormProviderSpec.scala (hunk 15 above)
Comment (aashik-varahan) [reply test/forms/RepaymentsContactEmailFormProviderSpec.scala]: Renamed to .contactEmail

File: test/forms/RepaymentsContactNameFormProviderSpec.scala (hunk 16 above)
Comment (aashik-varahan) [reply test/forms/RepaymentsContactNameFormProviderSpec.scala]: Renamed to .contactName

File: test/forms/RepaymentsTelephoneDetailsFormProviderSpec.scala (hunk 17 above)
Comment (aashik-varahan) [reply test/forms/RepaymentsTelephoneDetailsFormProviderSpec.scala]: Renamed to .telephoneNumber

File: test/controllers/repayments/RepaymentsTelephoneDetailsControllerSpec.scala (hunk 7 above)
Comment (aashik-varahan) [reply test/controllers/repayments/RepaymentsTelephoneDetailsControllerSpec.scala]: Renamed to Repayments Telephone Details Controller

File: test/navigation/RepaymentNavigatorSpec.scala (hunk 18 above)
Comment (aashik-varahan) [reply test/navigation/RepaymentNavigatorSpec.scala]: Renamed to reason for requesting a refund page

**Review by jassalrichy — DISMISSED** [review]

**Review by javed-iqbal1 — APPROVED** [review]

**Review by javed-iqbal1 — DISMISSED** [review]

**Review by javed-iqbal1 — APPROVED** [review]

**Pull request merged**